src/EnergyPlus/PoweredInductionUnits.cc

```diff
@@ -2069,6 +2115,418 @@ void CalcParallelPIU(EnergyPlusData &state,
+            thisPIU.heatingOperatingMode = HeatOpModeType::HeaterOff;
+            thisPIU.coolingOperatingMode = CoolOpModeType::CoolSecondStage;
+        } else {
+            // call regula falsi solver, vary a coooling control signal for fan speed and primary air flow together from min to max.
+            int constexpr MaxIte(500);    // Maximum number of iterations
+            Real64 constexpr Acc(0.0001); // Accuracy of result
+            int SolFla(0);                // Flag of solver
+
+            Real64 coolSignal = 0.5; // starting value
+            auto f = [&state, PIUNum, zoneLoad, ZoneNode](Real64 const coolSignal) {
+                return CalcVariableSpeedPIUCoolingResidual(state, coolSignal, PIUNum, zoneLoad, ZoneNode);
+            };
+
+            General::SolveRoot(state, Acc, MaxIte, SolFla, coolSignal, f, 0.0, 1.0);
+
+            if (SolFla == -1) {
+                ShowSevereError(state, "Iteration limit exceeded in calculating variable speed fan powered box cooling signal");
+                ShowContinueErrorTimeStamp(state, "");
+                ShowFatalError(state, format("Series PIU control failed for {}:{} ", thisPIU.UnitType, thisPIU.Name));
+            } else if (SolFla == -2) {
+                ShowSevereError(state, "Bad starting values for in calculating variable speed fan powered box cooling signal");
+                ShowContinueError(state, format("Zone Load to Cooling Setpoint = {:.2R} [W]", zoneLoad));
+                ShowContinueError(state, format("Load Delivered to Zone at Minimum Fan Speed  = {:.2R} [W]", qdotDelivMinPrim));
+                ShowContinueErrorTimeStamp(state, "");
+                ShowFatalError(state, format("Series PIU control failed for {}:{}", thisPIU.UnitType, thisPIU.Name));
+            } else {
+                thisPIU.PriAirMassFlow = coolSignal * (thisPIU.MaxPriAirMassFlow - thisPIU.MinPriAirMassFlow) +
+                                         thisPIU.MinPriAirMassFlow; // TODO: Check that this modification is correct
+                Real64 TotAirMassFlow = coolSignal * (thisPIU.MaxTotAirMassFlow - thisPIU.MinTotAirMassFlow) + thisPIU.MinTotAirMassFlow;
+                thisPIU.SecAirMassFlow = max(0.0, TotAirMassFlow - thisPIU.PriAirMassFlow);
+                thisPIU.heatingOperatingMode = HeatOpModeType::HeaterOff;
+                thisPIU.coolingOperatingMode = CoolOpModeType::CoolFirstStage;
+            }
+        }
+    }
+}
+
+void CalcVariableSpeedPIUStagedHeatingBehavior(EnergyPlusData &state,
+                                               int const PIUNum,   // number of the current PIU being simulated
+                                               int const ZoneNode, // zone node number
+                                               Real64 const zoneLoad,
+                                               bool const Pri,
+                                               Real64 const primaryAirMassFlow)
+{
```

Comment:
New function used the determine if, for staged heating operation, the variable speed fan should run at minimum, maximum, or in between the minimum and maximum to meet the load. If the latter, use regula falsi to find at which fraction it needs to run to meet the load. Note that this function is by both series and parallel PIUs.